lib/compress/zstd_fast.c

```diff
@@ -613,75 +633,200 @@ static size_t ZSTD_compressBlock_fast_extDict_generic(
-                U32 const offset = curr - matchIndex;
-                size_t mLength = ZSTD_count_2segments(ip+4, match+4, iend, matchEnd, prefixStart) + 4;
-                while (((ip>anchor) & (match>lowMatchPtr)) && (ip[-1] == match[-1])) { ip--; match--; mLength++; }   /* catch up */
-                offset_2 = offset_1; offset_1 = offset;  /* update offset history */
-                ZSTD_storeSeq(seqStore, (size_t)(ip-anchor), anchor, iend, OFFSET_TO_OFFBASE(offset), mLength);
-                ip += mLength;
-                anchor = ip;
-        }   }
-
-        if (ip <= ilimit) {
-            /* Fill Table */
-            hashTable[ZSTD_hashPtr(base+curr+2, hlog, mls)] = curr+2;
-            hashTable[ZSTD_hashPtr(ip-2, hlog, mls)] = (U32)(ip-2-base);
-            /* check immediate repcode */
-            while (ip <= ilimit) {
-                U32 const current2 = (U32)(ip-base);
-                U32 const repIndex2 = current2 - offset_2;
-                const BYTE* const repMatch2 = repIndex2 < prefixStartIndex ? dictBase + repIndex2 : base + repIndex2;
-                if ( (((U32)((prefixStartIndex-1) - repIndex2) >= 3) & (offset_2 <= curr - dictStartIndex))  /* intentional overflow */
-                   && (MEM_read32(repMatch2) == MEM_read32(ip)) ) {
-                    const BYTE* const repEnd2 = repIndex2 < prefixStartIndex ? dictEnd : iend;
-                    size_t const repLength2 = ZSTD_count_2segments(ip+4, repMatch2+4, iend, repEnd2, prefixStart) + 4;
-                    { U32 const tmpOffset = offset_2; offset_2 = offset_1; offset_1 = tmpOffset; }  /* swap offset_2 <=> offset_1 */
-                    ZSTD_storeSeq(seqStore, 0 /*litlen*/, anchor, iend, REPCODE1_TO_OFFBASE, repLength2);
-                    hashTable[ZSTD_hashPtr(ip, hlog, mls)] = current2;
-                    ip += repLength2;
-                    anchor = ip;
-                    continue;
-                }
-                break;
-    }   }   }
+            rval = MEM_read32(ip2) ^ 1; /* guaranteed to not match. */
+        }
+
+        /* write back hash table entry */
+        current0 = (U32)(ip0 - base);
+        hashTable[hash0] = current0;
+
+        /* check repcode at ip[2] */
+        if (MEM_read32(ip2) == rval) {
+            ip0 = ip2;
+            match0 = repBase + repIndex;
+            matchEnd = repIndex < prefixStartIndex ? dictEnd : iend;
+            assert((match0 != prefixStart) & (match0 != dictStart));
+            mLength = ip0[-1] == match0[-1];
+            ip0 -= mLength;
+            match0 -= mLength;
+            offcode = REPCODE1_TO_OFFBASE;
+            mLength += 4;
+            goto _match;
+        }
+
+        /* load match for ip[0] */
+        if (idx >= dictStartIndex) {
+            mval = MEM_read32(idxBase + idx);
+        } else {
+            mval = MEM_read32(ip0) ^ 1; /* guaranteed to not match. */
+        }
+
+        /* check match at ip[0] */
+        if (MEM_read32(ip0) == mval) {
```

Comment:
What about combining this branch with the previous one ?

Reply:
For reasons I didn't try to understand, I found that this construct was faster.

Reply:
I think it's because it gets assembled to something like:

```
int ival = MEM_read32(ip0);
int mval = ival ^ 1;
if (idx >= dictStartIndex) {
  mval = MEM_read32(idxBase + idx);
}
if (mval == ival) {
  goto _offset;
}
```

which I guess is pleasant for the CPU.

Reply:
Ok kind of pointless investigation but I was curious:

Original C:

```
/* load match for ip[0] */
if (idx >= prefixStartIndex) {
    mval = MEM_read32(base + idx);
} else {
    mval = MEM_read32(ip0) ^ 1; /* guaranteed to not match. */
}

/* check match at ip[0] */
if (MEM_read32(ip0) == mval) {
    /* found a match! */
    goto _offset;
}
```

`gcc-11 -O3 -DNDEBUG`:

```
mov    (%rdi),%edx        // edx /* ival */ = *rdi /* ip0 */;
mov    %edx,%ecx          // ecx /* mval */ = edx /* ival */;
xor    $0x1,%ecx          // ecx /* mval */ ^= 1;
cmp    -0x74(%rsp),%eax   // flags = cmp(prefixStartIndex, eax /* idx */);
jb     3935c              // if (flags.less_than) {
mov    %eax,%ecx          //   ecx /* mval */ = eax /* idx */;
mov    (%r14,%rcx,1),%ecx //   ecx /* mval */ = r14 /* base */[rcx /* mval */]; }
cmp    %edx,%ecx          // flags = cmp(edx /* ival */, ecx /* mval */);
je     393a1              // if (flags.equal) goto _offset;
```

Interpretation:

```
ival = *ip0;
mval = ival;
mval ^= 1;
if (prefixStartIndex < idx) {
  mval = idx;
  mval = base[mval];
}
if (ival == mval) goto _offset;
```

The `mval = idx` is weird... But ok.